File: src/main/java/com/ptaf/ai/context/SimilarFeatureRetriever.java
package com.ptaf.ai.context;

import com.ptaf.ai.config.AiAssistantProperties;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Locale;
import java.util.Set;
import java.util.regex.Pattern;

/**
 * Ranks existing feature snippets by simple keyword overlap with the requirement.
 */
public final class SimilarFeatureRetriever {
    private static final Pattern NON_WORD = Pattern.compile("[^a-z0-9\\s]");
    private static final Set<String> STOP_WORDS = Set.of(
            "the", "a", "an", "to", "of", "in", "on", "for", "with",
            "and", "or", "is", "are", "be", "should", "can", "user"
    );

    private final int maxSimilarFeatures;
    private final int minSimilarityScore;

    public SimilarFeatureRetriever(AiAssistantProperties properties) {
        this(properties.contextMaxSimilarFeatures(), properties.contextMinSimilarityScore());
    }

    public SimilarFeatureRetriever(int maxSimilarFeatures, int minSimilarityScore) {
        this.maxSimilarFeatures = Math.max(0, maxSimilarFeatures);
        this.minSimilarityScore = Math.max(0, minSimilarityScore);
    }

    public List<String> retrieve(String requirement, FrameworkGenerationContext context) {
        if (requirement == null || requirement.isBlank()) {
            return List.of();
        }
        if (context == null || context.existingFeatureSnippets() == null || context.existingFeatureSnippets().isEmpty()) {
            return List.of();
        }

        Set<String> reqTokens = tokenize(requirement);
        if (reqTokens.isEmpty()) {
            return List.of();
        }

        List<ScoredSnippet> ranked = new ArrayList<>();
        for (String snippet : context.existingFeatureSnippets()) {
            if (snippet == null || snippet.isBlank()) {
                continue;
            }
            Set<String> snippetTokens = tokenize(snippet);
            int score = overlapScore(reqTokens, snippetTokens);
            if (score >= minSimilarityScore) {
                ranked.add(new ScoredSnippet(snippet, score));
            }
        }

        ranked.sort(
                Comparator.comparingInt(ScoredSnippet::score).reversed()
                        .thenComparingInt(s -> s.snippet().length())
        );

        List<String> out = new ArrayList<>();
        for (ScoredSnippet s : ranked) {
            if (out.size() >= maxSimilarFeatures) {
                break;
            }
            out.add(s.snippet());
        }
        return out;
    }

    private static int overlapScore(Set<String> reqTokens, Set<String> snippetTokens) {
        int score = 0;
        for (String token : reqTokens) {
            if (snippetTokens.contains(token)) {
                score++;
            }
        }
        return score;
    }

    private static Set<String> tokenize(String text) {
        String normalized = NON_WORD.matcher(text.toLowerCase(Locale.ROOT)).replaceAll(" ");
        String[] parts = normalized.trim().split("\\s+");
        Set<String> tokens = new LinkedHashSet<>();
        for (String token : parts) {
            if (token.isBlank() || STOP_WORDS.contains(token)) {
                continue;
            }
            tokens.add(token);
        }
        return tokens;
    }

    private record ScoredSnippet(String snippet, int score) {
    }
}

File: src/main/ptaf/ai/context/similar_feature_retriever.py
from __future__ import annotations

import re

from ptaf.ai.config.ai_assistant_properties import AiAssistantProperties
from ptaf.ai.context.framework_generation_context import FrameworkGenerationContext

_NON_WORD = re.compile(r"[^a-z0-9\s]")
_STOP_WORDS = {
    "the", "a", "an", "to", "of", "in", "on", "for", "with",
    "and", "or", "is", "are", "be", "should", "can", "user",
}


class SimilarFeatureRetriever:
    def __init__(
        self,
        properties: AiAssistantProperties | None = None,
        *,
        max_similar_features: int | None = None,
        min_similarity_score: int | None = None,
    ) -> None:
        if properties is not None:
            self._max_similar_features = max(0, properties.context_max_similar_features())
            self._min_similarity_score = max(0, properties.context_min_similarity_score())
        else:
            self._max_similar_features = max(0, max_similar_features or 0)
            self._min_similarity_score = max(0, min_similarity_score or 0)

    def retrieve(self, requirement: str | None, context: FrameworkGenerationContext | None) -> list[str]:
        if not requirement or not requirement.strip():
            return []
        if not context or not context.existing_feature_snippets:
            return []

        req_tokens = _tokenize(requirement)
        if not req_tokens:
            return []

        ranked: list[tuple[str, int]] = []
        for snippet in context.existing_feature_snippets:
            if not snippet or not snippet.strip():
                continue
            score = _overlap_score(req_tokens, _tokenize(snippet))
            if score >= self._min_similarity_score:
                ranked.append((snippet, score))

        ranked.sort(key=lambda item: (-item[1], len(item[0])))
        return [snippet for snippet, _ in ranked[: self._max_similar_features]]


def _overlap_score(req_tokens: set[str], snippet_tokens: set[str]) -> int:
    return sum(1 for token in req_tokens if token in snippet_tokens)


def _tokenize(text: str) -> set[str]:
    normalized = _NON_WORD.sub(" ", text.lower())
    tokens: set[str] = set()
    for token in normalized.split():
        if token and token not in _STOP_WORDS:
            tokens.add(token)
    return tokens
